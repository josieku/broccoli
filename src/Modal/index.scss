.Modal-container {
    position: fixed;
    background: rgba(160, 160, 160, 0.5);
    height: 100%;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    z-index: 0;
    opacity: 1;
    animation: fadeIn 400ms;
    &.close {
        animation: fadeOut 400ms;
    }
    .Modal-box {
        background: white;
        width: 30vw;
        min-width: 300px;
        height: 55vh;
        min-height: 300px;
        max-height: 475px;
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        z-index: 1;
        h1 {
            margin-top: 0;
            font-size: 1.25em;
        }
        p {
            text-align: center;
            font-size: 0.9em;
            padding: 0 1.5em 0 1.5em;
        }
        .divider {
            height: 2px;
            width: 50px;
            background: black;
            margin: 0 0 2em 0;
        }
        .input-container {
            width: 100%;
            display: flex;
            justify-content: center;
            align-items: center;
            height: 15%;
            input:focus {
                border: 2px solid black;
            }
        }
        .box {
            font-family: SpaceMono;
            font-size: 0.9em;
            border: 1px solid black;
            padding: 0.5em;
            width: 85%;
            outline: none;
        }
        .input-container.error {
            height: 10%;
            align-items: center;
            p {
                font-style: italic;
                // font-size: 0.8em!important;
                // border: 1px dotted red;
                border-radius: 5px;
                max-width: 80%;
                padding: 0 1em 0 1em;
                margin: 0;
            }
        }
        button {
            box-sizing: content-box;
            margin: 2em 0 0 0 !important;
            cursor: pointer;
        }
        button:hover {
            background: lightgrey;
        }
        &.success {
            width: 40vw;
            height: 40vh;
        }
    }
}

@media (min-width: 450px) and (max-width: 600px) {
    .Modal-box h1 {
        font-size: 1em;
    }
    .Modal-box .box {
        font-size: 0.8em !important;
    }
    .Modal-box .input-container {
        height: 12% !important;
    }
    .Modal-box p {
        font-size: 0.8em !important;
    }
}

@media (min-width: 280px) and (max-width: 449px) {
    .Modal-box {
        height: 45vh !important;
        h1 {
            font-size: 0.9em;
        }
        .box {
            font-size: 0.7em !important;
        }
        .input-container {
            height: 12% !important;
        }
        p {
            font-size: 0.7em !important;
        }
    }
}

@media (max-width: 279px) {}

@keyframes fadeIn {
    0% {
        opacity: 0;
    }
    100% {
        opacity: 1;
    }
}

@keyframes fadeOut {
    0% {
        opacity: 1;
    }
    100% {
        opacity: 0;
    }
}