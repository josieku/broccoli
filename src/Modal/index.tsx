import React, { useState } from 'react';
import './index.scss';

const API_URL = "https://l94wc2001h.execute-api.ap-southeast-2.amazonaws.com/prod/fake-auth";

function Modal(props: { show: string; onClose: ()=>void }) {
    const [error, setError] = useState(""); // indicates input error
    const [success, setSuccess] = useState(false); // determines screen displayed
    const [isPosting, setIsPosting] = useState(false); // determines fetching state

    // input controls
    const [name, setName] = useState(""); 
    const [email, setEmail] = useState("");
    const [confirmEmail, setConfirmEmail] = useState("");

    // Clears all states (except isPosting).  Used for clearing input form/exiting modal
    const clearStates = () => {
        setError("");
        setSuccess(false);
        setName("");
        setEmail("");
        setConfirmEmail("");
    }

    // Triggers on button click, validates inputs & fires request
    const onClickSend = () => {
        if (name.length === 0 || email.length === 0 || confirmEmail.length === 0){
            return setError("Please fill in all fields");
        } else if (name.length <= 3){
            return setError("Full Name must have over 3 characters");
        } else if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)){
            return setError("Invalid email");
        } else if (email !== confirmEmail){
            return setError("Second email does not match the first");
        } else {
            return sendRequest();
        }
    }

    // Controls procedures when firing request to backend
    const sendRequest = async () => {
        setError(""); // clear previous error states
        setIsPosting(true); 

        await fetch(API_URL, { // fire post request with json body
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ name, email })
        })
        .then(response => response.json())
        .then(response=>{
            if (response.errorMessage) { // if there is an error
                setError("Error message from server");
            } else { // if error free, clear inputs and move to success screen
                clearStates();
                setSuccess(true);
            }
            setIsPosting(false);
        })
        .catch(() => {
            setError("Error message from server");
            setIsPosting(false);
        })
    }

    // Triggers when the modal background is clicked, or OK button on success screen
    const closeModal = async (e: React.MouseEvent) => {
        e.stopPropagation(); // stops event chain

        if (isPosting) return;

        props.onClose();
        setTimeout(clearStates, 500); // wait for animation to clear
    };

    return (
        <div 
            className={`Modal-container ${props.show}`} 
            style={props.show === "" ? { display: "none"} : {}}
            onClick={closeModal}
            >
            {
                success
                ? <SuccessScreen close={closeModal}/>
                : <RequestForm {...{
                    name, 
                    email, 
                    confirmEmail, 
                    isPosting,
                    error,
                    setName,
                    setEmail,
                    setConfirmEmail,
                    onClickSend
                }}/>
            }
        </div>
    );
}

interface RequestFormProps {
    name: string, 
    setName: (s: string)=>void,
    email: string, 
    setEmail: (s: string)=>void,
    confirmEmail: string, 
    setConfirmEmail: (s: string)=>void,
    onClickSend: ()=>void, 
    isPosting: boolean,
    error: string
}

function RequestForm(props: RequestFormProps) {
    return (
        <div className="Modal-box request" onClick={(e: React.MouseEvent<HTMLDivElement>) => e.stopPropagation()}>
                <h1>Request an invite</h1>
                <div className="divider"/>
                <div className="input-container">
                    <input 
                        className="box" 
                        placeholder="Full Name" 
                        value={props.name} 
                        disabled={props.isPosting}
                        onChange={(e:React.ChangeEvent<HTMLInputElement>)=>props.setName(e.target.value)}/>
                </div>
                <div className="input-container">
                    <input 
                        className="box" 
                        placeholder="Email" 
                        value={props.email}
                        disabled={props.isPosting}
                        onChange={(e:React.ChangeEvent<HTMLInputElement>)=>props.setEmail(e.target.value)}/>
                </div>
                <div className="input-container">
                    <input
                        className="box"
                        placeholder="Confirm Email"
                        value={props.confirmEmail}
                        disabled={props.isPosting}
                        onChange={(e:React.ChangeEvent<HTMLInputElement>)=>props.setConfirmEmail(e.target.value)}/>
                </div>
                <button 
                    className="box" 
                    onClick={props.onClickSend}
                    disabled={props.isPosting}>
                    {props.isPosting ? "Sending, please wait..." : "Send"}
                </button>
                <div className="input-container error">
                    {
                        props.error.length > 0 && !props.isPosting
                        ? <p>{props.error}</p>
                        : null
                    }
                </div>

            </div>
    )
}

function SuccessScreen(props: {close: (e:React.MouseEvent)=>void}) {
    return (
        <div className="Modal-box success" onClick={(e: React.MouseEvent<HTMLDivElement>) => e.stopPropagation()}>
                <h1>All done!</h1>
                <div className="divider"/>
                <p>You'll be one of the first to experience Broccoli {"&"} Co. when we launch.</p>
                <button 
                    className="box" 
                    onClick={(e: React.MouseEvent)=>props.close(e)}>
                    OK
                </button>
            </div>
    )
}


export default Modal;
