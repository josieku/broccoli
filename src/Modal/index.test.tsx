import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import {jest} from '@jest/globals';
import Modal from './index';

describe('renders the component', () => {
  it("matches snapshot", ()=> {
    const component = mount(<Modal show={"open"} onClose={()=>{}} />);
    expect(component).toMatchSnapshot();
  })
});

describe('input fields general validation', () => {
    let component:any;

    beforeEach(()=>{
        component = mount(<Modal show={"open"} onClose={()=>{}} />);
    })

    it("Name input changes", () => {
        component.find('input').at(0).simulate("change", { target: { value: 'Test' }});
        expect(component.find('input').at(0).props().value).toEqual('Test');
    })

    it("Email input changes", () => {
        component.find('input').at(1).simulate("change", { target: { value: 'Test' }});
        expect(component.find('input').at(1).props().value).toEqual('Test');
    })

    it("ConfirmEmail input changes", () => {
        component.find('input').at(2).simulate("change", { target: { value: 'Test' }});
        expect(component.find('input').at(2).props().value).toEqual('Test');
    })

    it("all fields must be filled", () => {
        expect(component.find('.input-container.error').children().length).toEqual(0);
        component.find('button.box').simulate('click');
        expect(component.find('.input-container.error').childAt(0).props().children).toEqual('Please fill in all fields');
    })
})

describe('input fields field specific validation', () => {
    let component:any;

    beforeEach(()=>{
        component = mount(<Modal show={"open"} onClose={()=>{}} />);
        component.find('input').at(0).simulate("change", { target: { value: 'test' }});
        component.find('input').at(1).simulate("change", { target: { value: 'test@test.com' }});
        component.find('input').at(2).simulate("change", { target: { value: 'test@test.com' }});
    })

    describe("Name Input", () => {
        it("Full name must have over 3 characters ('ff')", () => {
            component.find('input').at(0).simulate("change", { target: { value: 'ff' }});
            component.find('button.box').simulate('click');
            expect(component.find('.input-container.error').childAt(0).props().children).toEqual('Full Name must have over 3 characters');
        })

        it("Full name with over 3 characters work ('test')", () => {
            component.find('button.box').simulate('click');
            expect(component.find('button.box').props().children).toEqual('Sending, please wait...');
        })
    })

    describe("Email Input", ()=> {
        it("Email must be valid (ffff)", () => {
            component.find('input').at(1).simulate("change", { target: { value: 'ffff' }});
            component.find('button.box').simulate('click');
            expect(component.find('.input-container.error').childAt(0).props().children).toEqual('Invalid email');
        })

        it("Email must be valid (test@test@com)", () => {
            component.find('input').at(1).simulate("change", { target: { value: 'test@test@com' }});
            component.find('button.box').simulate('click');
            expect(component.find('.input-container.error').childAt(0).props().children).toEqual('Invalid email');
        })
    
        it("Email must be valid (test@.com)", () => {
            component.find('input').at(1).simulate("change", { target: { value: 'test@.com' }});
            component.find('button.box').simulate('click');
            expect(component.find('.input-container.error').childAt(0).props().children).toEqual('Invalid email');
        })
    })

    describe("Confirm Email Input", ()=>{
        it("Confirmation email must match email", () => {
            component.find('input').at(2).simulate("change", { target: { value: 'ff' }});
            component.find('button.box').simulate('click');
            expect(component.find('.input-container.error').childAt(0).props().children).toEqual('Second email does not match the first');
        })
    })

    describe("Request handling", () => {
        const mockSuccessFetchPromise = Promise.resolve({
            json: () => Promise.resolve("Registered"),
        });
        const mockFailFetchPromise: Promise<any> = Promise.resolve({ 
            json: () => Promise.resolve({errorMessage: "Bad Request"}),
        });

        const waitForComponentToFetch = async () => {
            await act(async () => {
                component.find('button.box').simulate('click');
                await new Promise(resolve => setTimeout(resolve));
                component.update();
            });
          };
        
        it("Failed request, shows error message", async () => {
            jest.spyOn(global, 'fetch').mockImplementation(():Promise<any> => mockFailFetchPromise);
            await waitForComponentToFetch();
            expect(component.find('.input-container.error').childAt(0).props().children).toEqual('Error message from server');        
        })

        it("Success request, moves onto success screen", async () => {
            jest.spyOn(global, 'fetch').mockImplementation(():Promise<any> => mockSuccessFetchPromise);
            expect(component.find('.Modal-box.success').length).toBe(0);
            await waitForComponentToFetch();
            expect(component.find('.Modal-box.success').length).toBe(1);
        })
    })

})