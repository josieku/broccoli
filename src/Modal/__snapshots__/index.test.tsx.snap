// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component matches snapshot 1`] = `
<Modal
  onClose={[Function]}
  show="open"
>
  <div
    className="Modal-container open"
    onClick={[Function]}
    style={Object {}}
  >
    <RequestForm
      confirmEmail=""
      email=""
      error=""
      isPosting={false}
      name=""
      onClickSend={[Function]}
      setConfirmEmail={[Function]}
      setEmail={[Function]}
      setName={[Function]}
    >
      <div
        className="Modal-box request"
        onClick={[Function]}
      >
        <h1>
          Request an invite
        </h1>
        <div
          className="divider"
        />
        <div
          className="input-container"
        >
          <input
            className="box"
            disabled={false}
            onChange={[Function]}
            placeholder="Full Name"
            value=""
          />
        </div>
        <div
          className="input-container"
        >
          <input
            className="box"
            disabled={false}
            onChange={[Function]}
            placeholder="Email"
            value=""
          />
        </div>
        <div
          className="input-container"
        >
          <input
            className="box"
            disabled={false}
            onChange={[Function]}
            placeholder="Confirm Email"
            value=""
          />
        </div>
        <button
          className="box"
          disabled={false}
          onClick={[Function]}
        >
          Send
        </button>
        <div
          className="input-container error"
        />
      </div>
    </RequestForm>
  </div>
</Modal>
`;
